### ISSUE TITLE
new rule - detect when decorators are in the wrong order

### ISSUE THREAD

**Issue by DetachHead** [astral-sh/ruff#6697]
some examples:
```py
from collections.abc import Iterator
from contextlib import contextmanager


class Foo:
    @contextmanager
    @staticmethod
    def foo() -> Iterator[None]:
        yield None

    def bar(self):
        with self.foo():
            ...


Foo().bar() # TypeError: Foo.foo() takes 0 positional arguments but 1 was given
```
```py
from abc import abstractmethod

class Foo:
    @abstractmethod # AttributeError: attribute '__isabstractmethod__' of 'property' objects is not writable
    @property
    def foo(self) -> int:
        ...
```
both of these errors are fixed by swapping the order of the decorators

**Comment by charliermarsh** [comment]
Would probably require that we maintain a set of known, proper orderings for the standard library decorators (like `@property`, etc.), but that's not too hard...

**Comment by namurphy** [comment]
Thank you for raising this issue!

I'm wondering if this rule could have a configuration option so that users could add their own decorator ordering rules.  Even if not needed for correctness, this could help improve consistency and readability.

There's also a chance that users might want to specify the order of ≥3 decorators...though this would quickly add complexity to the rule.  For ≥3 stacked decorators, a possible edge case would be when two decorators have to be in a certain order, but there might be a middle intervening decorator:

```python
from abc import abstractmethod

class Foo:
    @abstractmethod
    @here_be_dragons
    @property
    def foo(self) -> int:
        ...
```

I don't know if this happens often in the pythoniverse (hopefully not!), but I'm not sure what would be best to do in this case.  Perhaps it would be best for the rule to not be applied to cases with an intervening decorator.

**Comment by zanieb** [comment]
I think adding configuration should generally be carefully considered only once we have a reasonable first implementation. I often get confused about the correct order of property / method decorators as well, I think this would be pretty useful even in a limited scope.

I'm not sure we should handle cases with interleaved decorators.

**Comment by charliermarsh** [comment]
I like the idea of handling the known, common, invalid permutations from the standard library.

**Comment by namurphy** [comment]
Agreed on all of the above!  Perhaps my suggestion for a configuration option for this rule would be best brought up in a follow-up issue after this rule gets implemented for decorators in the standard library.  

For anyone curious about the use case motivating my suggestions, we use two decorators in PlasmaPy to process and validate arguments to functions and methods.  The order of the decorators matters when they are applied to instance or class methods.  When applied to class methods (which is admittedly rare), the `@classmethod` decorator would also be present. Right now, we handle the situation by noting this in admonitions in the docstring of [`@particle_input`](https://docs.plasmapy.org/en/stable/api/plasmapy.particles.decorators.particle_input.html#particle-input).  This use case motivates a configuration option, but I have no idea if it's a common enough use case to warrant adding the capability.

Thank you!

**Comment by evanrittenhouse** [comment]
Feel free to assign this to me, I can take a look

**Comment by zanieb** [comment]
@evanrittenhouse basically by using them and seeing if they behave correctly at runtime.

The potentially relevant ones I can think of are:
- `abc.abstractmethod`
- `classmethod`
- `staticmethod`
- `property`
- `contextlib.contextmanager`
- `contextlib.asyncontextmanager`
- `functools.wraps`
- `functools.cache`
- `functools.cached_property` ([maybe helpful context](https://docs.python.org/3/faq/programming.html#faq-cache-method-calls))
- `functools.lru_cache`

**Comment by DetachHead** [comment]
also `property.setter`

**Comment by evanrittenhouse** [comment]
@zanieb Thanks for the list - was hoping someone knew of some docs somewhere, but off to REPL I go...

**Comment by DetachHead** [comment]
also `@final` and `@deprecated`

for some reason it's apparently intended behavior for `@deprecated` to only work if there are no other decorators under it (https://github.com/microsoft/pylance-release/issues/4600#issuecomment-1632023633)

**Comment by irdkwmnsb** [comment]
I'm using ruff for a project of mine and wanted to add custom lint rules for my own decorators.  
If this feature gets implemented I would also really love the ability to specify my own order for the decorators.
For example when using with https://python-dependency-injector.ets-labs.org/wiring.html#decorator-inject

**Comment by anishgirianish** [comment]
Hi, I would like to work on this if this is still open. Thanks

**Comment by anishgirianish** [comment]
Hi! I've opened a PR for this: [#23461](https://github.com/astral-sh/ruff/pull/23461)                                                                                                                            
                                                          
Would love a review when you get a chance. Happy to adjust based on feedback!
